### ISSUE TITLE
Scatter plot render artifacts

### ISSUE THREAD

**Issue by grabul** [ScottPlot/ScottPlot#889]
If you have different densities on the x-axis of your scatter data and you zoom into a high density area there are render errors drawing additional horizontal lines where we shouldn't have there.
example code:

```cs
var plt = new ScottPlot.Plot(600, 400);
int pointCount = 1000;
double[] ys = new double[pointCount];
double[] xs = new double[pointCount];
double currentX = 0;
 for (int i = 0; i < pointCount; i++)
 {
    if ((i % 100) < 10)
       currentX += 10;
    else
       currentX += 0.001;
    xs[i] = currentX;
    ys[i] = i % 2 == 0 ? 1.0 : 0.0;
 }
plt.AddScatterStep(xs, ys);
```

Also belongs to a normal ScatterPlot (no step style) where you add additional data points to create the steps "by hand" within the loop like:

```cs
var plt = new ScottPlot.Plot(600, 400);
int pointCount = 1000;
var xs = new List<double>();
var ys = new List<double>();
double last_y = -1.0;
double currentX = 0;
for (int i = 0; i < pointCount; i++)
{
  if ((i % 100) < 10)
     currentX += 10;
  else
     currentX += 0.001;

  if (last_y >= 0.0)
  {
     xs.Add(currentX);
     ys.Add(last_y);
  }

  xs.Add(currentX);
  last_y = i % 2 == 0 ? 1.0 : 0.0;
  ys.Add(last_y);
}
plt.AddScatter(xs.ToArray(), ys.ToArray());
```

![scatter](https://user-images.githubusercontent.com/80955161/111768050-b6a33f80-88a7-11eb-8f90-7946649ba3d0.jpg)

**Comment by swharden** [comment]
Hi @grabul, thanks for reporting this!

## The Problem is GDI+

This looks like a rendering bug related to GDI+. It occurs when lines are drawn that extend extremely long distances off the screen.

![artifact](https://user-images.githubusercontent.com/4165489/111889958-7d5bf400-89bb-11eb-97df-6ea4b024a821.gif)

## Immediate Fix / Workaround

An immediate fix would be for the user to dice-up their data a bit more to ensure there aren't data points with extremely long gaps. It's not elegant, but it would work right away.

## Best Solution

A true fix would be to check the first and last point right before rendering to see if it extends extremely far off the screen and if it does determine the slope from the first visible point and replace each with a shorter line that extends to just beyond the edge of the plot. This is a pretty niche case so I'm tempted to prioritize some of the larger issues short-term (#716), but I'll leave this issue open until someone else or I has time to consider implementing a solution.

**Comment by grabul** [comment]
Thanks for the fast feedback!

Well I'm not quite sure if this belongs to the gdi rendering, because if you change in the second code snippet the AddScatter to AddSignalXY then you don't have that issue at all.
But then you will get a new render issue of the SignalPlotXY now: At the right side of the plot area you can see that during zoom the next point to draw the line to sometimes is not correct. It seems to me the problem belongs to that two consecutive points have the same x-value and the binary search misses the vertical point and hits the point after that. This problem exits also in the area where you don't have a high density of data point.

https://user-images.githubusercontent.com/80955161/111906741-6e745080-8a52-11eb-8f72-da5424f5f7fe.mp4

**Comment by StendProg** [comment]
Hi @grabul, @swharden 

> Well I'm not quite sure if this belongs to the gdi rendering, because if you change in the second code snippet the AddScatter to AddSignalXY then you don't have that issue at all.

It's GDI+ clipping problem like @swharden described.
SignalXY implements it's own clipping to prevent this effect.
Thanks's for noting `SignalXY` right edge bug. I think it's just some logic bug, because left border uses the same alg and have no problem. I'll see what can be done with this in the week.

> A true fix would be to check the first and last point right before rendering to see if it extends extremely far off the screen and if it does determine the slope from the first visible point and replace each with a shorter line that extends to just beyond the edge of the plot. This is a pretty niche case so I'm tempted to prioritize some of the larger issues short-term (#716), but I'll leave this issue open until someone else or I has time to consider implementing a solution.

For a `Scatter` with its points in general not ordered at all, I'm afraid we need to check every pair of points, and the cutting algorithms can turn out to be quite complicated.

**Comment by swharden** [comment]
Summarizing conclusions once more:
* This is a GDI+ clipping problem which cannot be easily fixed
* There was a bug in SignalXY, but it was fixed by #893 (I'll release an updated NuGet package later today)
